### PR TITLE
feat(ast) Port arrayjoin optimnizations to the AST

### PR DESCRIPTION
**Pull request by fpacifici**
This ports the logic of this file that optimizes arrayJoin clauses: https://github.com/getsentry/snuba/blob/master/snuba/datasets/tags_column_processor.py

Unfortunately the test specifically on the AST is a lot more verbose.

Introduces the ArrayJoin optimizer that applies these optimizations:
- if tags_key and tags_value are both present in the query, instead of running multiple arrayJoins, it runs a single arrayJoin of the zip of the tags.key and tags.value column.
- if there are conditions based on the tags_key field in the query, we filter those keys with arrayFilter before doing the arrayJoin instead of arrayjoinning and then use the where and having conditions to do the filter.

#983 depends on this and wires the query processor up.

### CODE REVIEW

File: snuba/query/conditions.py [hunk 1]
```diff
@@ -66,6 +68,19 @@ def __is_set_condition(exp: Expression, operator: str) -> bool:
...
+        None,
+        String(operator),
+        (
+            Param("lhs", lhs),
+            Param("tuple", FunctionCallPattern(None, String("tuple"), None)),
+        ),
+    )
+
+
 def in_condition(
```
Comment (fpacifici) [inline snuba/query/conditions.py:84]: I think at a point we may remove the old helper methods (is_in_condition, is_binary_condition) and just replace all of them with the Patterns.

File: snuba/query/processors/arrayjoin_optimizer.py [hunk 2]
```diff
@@ -0,0 +1,288 @@
...
+from snuba.query.expressions import Literal as LiteralExpr
+from snuba.query.matchers import Any, Column, FunctionCall, Literal, Or, Param, String
+from snuba.request.request_settings import RequestSettings
+
+tags_key_pattern = FunctionCall(
+    None, String("arrayJoin"), (Column(None, None, String("tags.key")),)
+)
+
+
+def _get_tag_keys_in_condition(condition: Optional[Expression]) -> Optional[Set[str]]:
```
Comment (fpacifici) [inline snuba/query/processors/arrayjoin_optimizer.py:28]: Does the same as this:
https://github.com/getsentry/snuba/blob/master/snuba/datasets/tags_column_processor.py#L160

I decided not to try to make the old logic parametric and to just start fresh instead because, specifically in the query processor code the structure is fairly different because we have the matcher api right now.

File: snuba/query/processors/arrayjoin_optimizer.py [hunk 3]
```diff
@@ -0,0 +1,288 @@
...
+                # it is tags_key since we found filtered tag keys.
+                return _filtered_tag_keys(expr.alias, filtered_tags)
+            else:
+                # No viable optimization
+                return expr
+
+        query.transform_expressions(replace_expression)
+
+
+def _unfiltered_tag_pairs(alias: Optional[str], array_pos: LiteralExpr) -> Expression:
```
Comment (fpacifici) [inline snuba/query/processors/arrayjoin_optimizer.py:173]: all the methods below are introduced to make the expressions simpler to understand and disect even if some times they are used only once.
Expressions like this:
```
(arrayJoin(arrayFilter(
    #       pair -> arrayElement(pair, 1) IN (tags),
    #       arrayMap((x,y) -> [x,y], tags.key, tags.value)
    #  )) as all_tags)[1]
```
in the ast are gigantic and they can get to 20 levels of nesting, which makes them very hard to understand.

File: tests/query/test_conditions.py [hunk 4]
```diff
@@ -16,7 +19,7 @@ def test_expressions_from_basic_condition() -> None:
     """
 
     c = Column(None, "t1", "c1")
-    f1 = FunctionCall(None, "f", [c])
+    f1 = FunctionCall(None, "f", (c,))
```
Comment (fpacifici) [inline tests/query/test_conditions.py:22]: old mistake from before we turned sequences into tuples.

File: snuba/query/processors/arrayjoin_optimizer.py [hunk 5]
```diff
@@ -0,0 +1,300 @@
...
+    ast_condition = query.get_condition_from_ast()
+    cond_tags_key = (
+        _get_tag_keys_in_condition(ast_condition)
+        if ast_condition is not None
+        else set()
+    )
+    if cond_tags_key is None:
+        # This means we found an OR. Cowardly we give up even though there could
+        # be cases where this condition is still optimizable.
+        return set()
```
Comment (lynnagara) [inline snuba/query/processors/arrayjoin_optimizer.py:95]: I don't quite get why we return the empty set here and totally skip the attempt to perform the optimization on HAVING. And vice versa. Wouldn't that still be valid / unrelated to whether or not we found an OR here?

File: snuba/query/processors/arrayjoin_optimizer.py [hunk 6]
```diff
@@ -0,0 +1,300 @@
...
+        _get_tag_keys_in_condition(ast_having) if ast_having is not None else set()
+    )
+    if having_tags_key is None:
+        # Same as above
+        return set()
+
+    return cond_tags_key | having_tags_key
+
+
+class ArrayjoinOptimizer(QueryProcessor):
```
Comment (lynnagara) [inline snuba/query/processors/arrayjoin_optimizer.py:108]: It's not used anywhere, is it supposed to be applied to the AST already?

File: snuba/query/processors/arrayjoin_optimizer.py [hunk 7]
```diff
@@ -0,0 +1,300 @@
...
+    array_index: LiteralExpr,
+) -> Expression:
+    # (arrayJoin(arrayFilter(
+    #       pair -> arrayElement(pair, 1) IN (tags),
+    #       arrayMap((x,y) -> [x,y], tags.key, tags.value)
+    #  )) as all_tags)[1]
+    return arrayElement(
+        alias,
+        array_join(
+            "all_tags",
```
Comment (lynnagara) [inline snuba/query/processors/arrayjoin_optimizer.py:209]: What happens if someone actually uses the alias "all_tags" in the query?

Comment (fpacifici) [reply snuba/query/processors/arrayjoin_optimizer.py:209]: Sadly the same as what happens today: https://github.com/getsentry/snuba/blob/master/snuba/datasets/tags_column_processor.py#L306-L309
This same alias is already being applied in the processing code today. You would have alias shadowing.

I think, now with the ast, we could ensure the alias is not already used before when applying it.

File: snuba/query/processors/arrayjoin_optimizer.py (hunk 5 above)
Comment (fpacifici) [reply snuba/query/processors/arrayjoin_optimizer.py:95]: We cannot guarantee the query would be correct.
This is the same logic we execute today:
https://github.com/getsentry/snuba/blob/master/snuba/datasets/tags_column_processor.py#L243-L245

The issue is that if we are using tags_key and tags_value together in the same query, the arrayJoin would happen only once and then we would use the alias in the other cases. If the `WHERE` condition can be optimized but the `HAVING` cannot, and we optimized the `WHERE`, then we would end up using the same condition in the having (the `all_tags` alias), which would be wrong.

So this takes the conservative approach and just bail in case it is not 100% sure the condition is optimizable at the cost of having false negatives (conditions that could be optimized but that are not).

File: snuba/query/processors/arrayjoin_optimizer.py (hunk 6 above)
Comment (fpacifici) [reply snuba/query/processors/arrayjoin_optimizer.py:108]: Sorry forgot to mention it:
it is wired up here https://github.com/getsentry/snuba/pull/983

File: snuba/query/processors/arrayjoin_optimizer.py (hunk 7 above)
Comment (lynnagara) [reply snuba/query/processors/arrayjoin_optimizer.py:209]: Sure. I think adding that check before picking a name, adding some logging, or even just renaming to `__all_tags` (or something that makes the possibility of collision clearer) would all be fine.

Comment (tkaemming) [reply snuba/query/processors/arrayjoin_optimizer.py:209]: Reserving a `snuba:` namespace these sorts of optimizations might not be a bad idea.

**Review by tkaemming — APPROVED** [review]

File: snuba/query/processors/arrayjoin_optimizer.py [hunk 8]
```diff
@@ -0,0 +1,300 @@
...
+from snuba.query.expressions import Column as ColumnExpr
+from snuba.query.expressions import Expression
+from snuba.query.expressions import FunctionCall as FunctionCallExpr
+from snuba.query.expressions import Lambda
+from snuba.query.expressions import Literal as LiteralExpr
+from snuba.query.matchers import Any, Column, FunctionCall, Literal, Or, Param, String
+from snuba.request.request_settings import RequestSettings
+
+tags_key_pattern = FunctionCall(
+    None, String("arrayJoin"), (Column(None, None, String("tags.key")),)
```
[the remark this answers is not in the document]
Comment (fpacifici) [reply snuba/query/processors/arrayjoin_optimizer.py:24]: Do you mean the behavior of matching a column present in an arrayJoin function call ?

File: snuba/query/processors/arrayjoin_optimizer.py [hunk 9]
```diff
@@ -0,0 +1,300 @@
...
+)
```
[the remark this answers is not in the document]
Comment (fpacifici) [reply snuba/query/processors/arrayjoin_optimizer.py:25]: Ah yes, this was written before I addressed that one. Will fix.

File: snuba/query/processors/arrayjoin_optimizer.py [hunk 10]
```diff
@@ -0,0 +1,300 @@
...
+    #   tag -> tag IN (tags),
+    #   tags.key
+    # ))
+    return array_join(
+        alias, filter_keys(ColumnExpr(None, None, "tags.key"), filtered_tags),
+    )
+
+
+def array_join(alias: Optional[str], content: Expression) -> Expression:
+    return FunctionCallExpr(alias, "arrayJoin", (content,))
```
[the remark this answers is not in the document]
Comment (fpacifici) [reply snuba/query/processors/arrayjoin_optimizer.py:235]: I generally did not put them there at first, because they are used only in one place now and I did not really want to encourage people to use arrayJoin, but very weak opinion.

File: snuba/query/processors/arrayjoin_optimizer.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (tkaemming) [reply snuba/query/processors/arrayjoin_optimizer.py:24]: It seems possible (maybe not likely, but possible) that any mapping-like `Nested` column (e.g. `contexts.*`) could have an `arrayJoin` applied to it which could be optimized in the same fashion. Or am I overlooking something, and this truly unique to the concept of tags and not generically any key/value mapping?

[the remark this answers is not in the document]
Comment (fpacifici) [reply snuba/query/processors/arrayjoin_optimizer.py:24]: This can in principle be applied to any key-value pair column. As a matter of fact I am still not sure why this was not applied to contexts to begin with (https://github.com/getsentry/snuba/blob/master/snuba/datasets/tags_column_processor.py#L111).
I think one of the reasons was that (fortunately) there is no such a thing as a `contexts_key` and `conntexts_value` keywords, still that did not prevent the user from actually send a valid `arrayJoin(contexts.key)` function, so that does not explain this enttirely.

I think the question is whether we want to encourage using `arrayJoin` at all.

For this PR I can make the column name parametric, though I would not go as far as wiring this processor for contexts as well unless we knew if there was a reason for not doing so.

[the remark this answers is not in the document]
Comment (tkaemming) [reply snuba/query/processors/arrayjoin_optimizer.py:24]: > I think the question is whether we want to encourage using `arrayJoin` at all.

I don't have a super strong opinion about `arrayJoin` in isolation. Personally, I'd rather encourage using `arrayJoin` directly if the only other option is `tags_{key,value}` since the behavior is at least a bit more obvious or explainable.

> For this PR I can make the column name parametric, 

I think this a good idea, looking at #983, I think the fact that there is an implicit schema requirement between the processor and the storage is problematic, regardless of whether or not other columns like `contexts` are included in the processor's responsibilities.

> though I would not go as far as wiring this processor for contexts as well unless we knew if there was a reason for not doing so.

I'm not sure how to parse this triple negative with much confidence I'm interpreting the intent correctly, can you rephrase this?

[the remark this answers is not in the document]
Comment (fpacifici) [reply snuba/query/processors/arrayjoin_optimizer.py:24]: "I'm not sure how to parse this triple negative with much confidence I'm interpreting the intent correctly, can you rephrase this?"

I would not apply this optimization to contexts just yet. None would be using it, and I am not 100% sure it would work just fine. I'd like to figure out why contexts were not supported i nthe first place.

[the remark this answers is not in the document]
Comment (tkaemming) [reply snuba/query/processors/arrayjoin_optimizer.py:24]: Ok, I suppose this makes sense, thanks for clarifying.

File: snuba/query/processors/arrayjoin_optimizer.py (hunk 7 above)
Comment (fpacifici) [reply snuba/query/processors/arrayjoin_optimizer.py]: added the prefix, will put something on the backlog to enforce `snuba:` cannot be used by the client

File: snuba/query/processors/arrayjoin_optimizer.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (fpacifici) [reply snuba/query/processors/arrayjoin_optimizer.py]: made it parametric

File: snuba/query/processors/arrayjoin_optimizer.py [hunk 11]
```diff
@@ -0,0 +1,300 @@
...
+            None,
+            String(ConditionFunctions.EQ),
+            (tags_key_pattern, Literal(None, Param("key", Any(str)))),
+        ).match(c)
+        if match is not None:
+            tag_keys_found.add(match.string("key"))
+
+        match = is_in_condition_pattern(tags_key_pattern).match(c)
+        if match is not None:
+            function = cast(FunctionCallExpr, match.expression("tuple"))
```
[the remark this answers is not in the document]
Comment (fpacifici) [reply snuba/query/processors/arrayjoin_optimizer.py]: I was trying to do this in one statement instead of two, but I realized that a cast does not enforce, so turned it into assert.

File: snuba/query/processors/arrayjoin_optimizer.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (fpacifici) [reply snuba/query/processors/arrayjoin_optimizer.py]: I moved arrayJoin

File: snuba/query/processors/arrayjoin_optimizer.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (fpacifici) [reply snuba/query/processors/arrayjoin_optimizer.py]: Turned it into ArrayJoinOptimizer. 
Do you mean making the name more descriptive? Because I have a very hard time describing the idea of joining two arrayjoin into one in a name.
Maybe ArrayJoinReducer ?

File: snuba/query/processors/arrayjoin_optimizer.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (tkaemming) [reply snuba/query/processors/arrayjoin_optimizer.py]: TBH, I'm not really sure what the use case for `cast` is. It seems like in most cases asserting on `isinstance` is safer, and assertions can be disabled with the `-O` flag anyway if performance is a priority.

File: snuba/query/processors/arrayjoin_optimizer.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (tkaemming) [reply snuba/query/processors/arrayjoin_optimizer.py]: I think using `Optimizer` in the name is fine but clarifying that this works on array joins of a specific form (mapping-style nested field with associated key, value pairs) would make this a bit more straightforward without having to read the docstring. Otherwise out of context this kind of implies it just works on array joins broadly which is not quite true.

File: snuba/query/processors/arrayjoin_optimizer.py [hunk 12]
```diff
@@ -263,26 +324,27 @@ def filter_key_values(
...
             key_values,
         ),
     )
 
 
-def filter_keys(keys: Expression, tag_keys: Sequence[LiteralExpr]) -> Expression:
+def filter_keys(column: Expression, keys: Sequence[LiteralExpr]) -> Expression:
     """
-    Filter a an array of tag keys based on a sequence of tag keys.
+    Filter a a Column array based on a sequence of keys.
```
Comment (tkaemming) [inline snuba/query/processors/arrayjoin_optimizer.py:341]:
```suggestion
    Filter a Column array based on a sequence of keys.
```

**Review by tkaemming — APPROVED** [review]

File: snuba/query/processors/arrayjoin_optimizer.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (fpacifici) [reply snuba/query/processors/arrayjoin_optimizer.py]: ahhh ok, sorry I misunderstood. Will revert and add some context.
